### PR TITLE
Initial changes for serialization helpers

### PR DESCRIPTION
**Pull request by m-nash**
Fixes https://github.com/microsoft/typespec/issues/3580
Fixes https://github.com/microsoft/typespec/issues/3536

### CODE REVIEW

File: packages/http-client-csharp/generator/TestProjects/Local/Unbranded-TypeSpec/src/Generated/Internal/ModelSerializationExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+        public static string GetRequiredString(this System.Text.Json.JsonElement element)
+        {
+            var value = element.GetString();
+            if (value == null)
+            {
+                throw new InvalidOperationException($"The requested operation requires an element of type 'String', but the target element has type '{element.ValueKind}'.");
+            }
+            return value;
+        }
+
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/TestProjects/Local/Unbranded-TypeSpec/src/Generated/Internal/ModelSerializationExtensions.cs:125]: Should we generate minimal description?

File: packages/http-client-csharp/generator/TestProjects/Local/Unbranded-TypeSpec/src/Generated/Internal/ModelSerializationExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+        /// <param name="writer"></param>
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/TestProjects/Local/Unbranded-TypeSpec/src/Generated/Internal/ModelSerializationExtensions.cs:126]: nit: should we fill these out?

File: packages/http-client-csharp/generator/TestProjects/Local/Unbranded-TypeSpec/src/Generated/Internal/ModelSerializationExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+    {
+        internal static readonly System.ClientModel.Primitives.ModelReaderWriterOptions WireOptions = new System.ClientModel.Primitives.ModelReaderWriterOptions("W");
+
+        /// <param name="element"></param>
+        public static object GetObject(this System.Text.Json.JsonElement element)
+        {
+            switch (element.ValueKind)
+            {
+            case System.Text.Json.JsonValueKind.String: 
+            return element.GetString();
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/TestProjects/Local/Unbranded-TypeSpec/src/Generated/Internal/ModelSerializationExtensions.cs:24]: nit: indentation

File: packages/http-client-csharp/generator/TestProjects/Local/Unbranded-TypeSpec/src/Generated/Internal/ModelSerializationExtensions.cs (hunk 1 above)
Comment (m-nash) [reply packages/http-client-csharp/generator/TestProjects/Local/Unbranded-TypeSpec/src/Generated/Internal/ModelSerializationExtensions.cs:125]: These are internal so not needed but if you think its a custom code usability thing lets open a tracking issue and we can address it when we do the custom code work.

File: packages/http-client-csharp/generator/TestProjects/Local/Unbranded-TypeSpec/src/Generated/Internal/ModelSerializationExtensions.cs (hunk 2 above)
Comment (m-nash) [reply packages/http-client-csharp/generator/TestProjects/Local/Unbranded-TypeSpec/src/Generated/Internal/ModelSerializationExtensions.cs:126]: removed for internal methods now

**Review by JoshLove-msft — APPROVED** [review]

File: cspell.yaml [hunk 4]
```diff
@@ -146,6 +146,15 @@ ignorePaths:
   - .editorconfig
   - .github/CODEOWNERS
   - packages/samples/test/output/**
+  - "**/BenchmarkDotNet.Artifacts/**"
+  - cspell.yaml
+overrides:
+  - filename: "packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/test/TypeFormatterTests.cs"
```
Comment (timotheeguerin) [inline cspell.yaml:152]: can you not have `cspell:ignore` in comment instead?

https://cspell.org/configuration/document-settings/#enable--disable-checking-sections-of-code

Comment (m-nash) [reply cspell.yaml:152]: merged before I saw this https://github.com/microsoft/typespec/pull/3607

**Pull request merged**